aws-connect

```diff
@@ -208,6 +224,70 @@ elif [ "${action}" == "tunnel" ]; then
     --parameters "{\"portNumber\":[\"22\"],\"localPortNumber\":[\"${port}\"]}" \
     --region "${aws_region}" \
     --profile "${aws_profile}"
+elif [ "${action}" == "document" ]; then
+
+  if [ -z "${github_token_location}" ]
+  then
+    echo "No github token provided - scripts referenced in SSM document must be in a public repo if no token is provided"
+  else
+    arguments="githubTokenLocation='${github_token_location}',"
+  fi
+
+  if [ -z "${document_parameters}" ]
+  then
+    arguments="${arguments::${#arguments}-1}"
+  else
+    arguments="${arguments}parameters='${document_parameters}'"
+  fi
+
+  if [ -z "${cloudwatch_group}" ]
+  then
+    cloudwatch_group="aws-connect"
+  fi
+
+  echo "Running ${document_name} on instance ${instance_id}"
+
+  sh_command_id=$( aws ssm send-command \
+    --instance-ids ${instance_id} \
+    --document-name "${document_name}" \
+    --comment "Document run using aws-connect" \
+    --output text \
+    --parameters "${arguments}" \
+    --region ${aws_region} \
+    --profile ${aws_profile} \
+    --cloud-watch-output-config '{"CloudWatchLogGroupName":"'${cloudwatch_group}'","CloudWatchOutputEnabled":true}' \
+    --query "Command.CommandId")
+
+  status=$(aws ssm list-command-invocations \
+    --command-id "${sh_command_id}" \
+    --details \
+    --output text \
+    --no-paginate \
+    --query "CommandInvocations[0].Status")
+
+  echo "Command ID: $sh_command_id"
+
+  # wait for the command to finish running before getting the output
+  while [ "$status" == "InProgress" ]
+  do
+    echo "Command still running..."
+    sleep 5
+
+    status=$(aws ssm list-command-invocations \
+      --command-id "${sh_command_id}" \
+      --details \
+      --output text \
+      --no-paginate \
+      --query "CommandInvocations[0].Status")  
+  done
+
+  if [ "$status" == "Failed" ]
+  then
+    echo "Command Failed!"
+  fi
+
+  printf "$(aws logs get-log-events --log-group-name ${cloudwatch_group} --log-stream-name ${sh_command_id}/${instance_id}/runShellScript/stdout --query "events[0].message")"
```

Comment:
2 comments:

1) For the log stream name, why does it contain `runShellScript`?  Isn't that the name of AWS's SSM document?  It's probably better to include the document name being run AND the user running it (since the log group is shared).  There's a CLI call to get the current IAM user

2) You probably read this, but this is paginated

>By default, this operation returns as many log events as can fit in a response size of 1MB (up to 10,000 log events). You can get additional log events by specifying one of the tokens in a subsequent call.

It's probably ok to just print a line showing the full URL to the log stream for the full logs.  Something like

> Log output is limited to 1MB.  For the full logs, view them at https://console.aws.com....blah blah/log_group/stream (look at an existing for the full url)